### PR TITLE
Fix IPC timeout issue on APL&CNL 

### PR DESCRIPTION
**Pull request by RanderWang**
DSP only processes one IPC msg at one time, so driver should do as it required. If two IPC msgs are sent to DSP at the same time, IPC timeout. Now check is_dsp_ready at the IPC sent time to comply with the rule.

Also Refine the is_dsp_ready function, check both BUSY bit and DONE bit. To make sure the last IPC msg is not overwriten, DONE bit is set in cmd_done function which is called at the end of msg processing function.

### CODE REVIEW

File: sound/soc/sof/intel/cnl.c [hunk 1]
```diff
@@ -147,10 +147,12 @@ static int cnl_ipc_cmd_done(struct snd_sof_dev *sdev, int dir)
...
 {
-	u64 val;
+	u64 busy, done;
 
-	val = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR);
-	if (val & CNL_DSP_REG_HIPCIDR_BUSY)
+	busy = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDR);
+	done = snd_sof_dsp_read(sdev, HDA_DSP_BAR, CNL_DSP_REG_HIPCIDA);
+	if ((busy & CNL_DSP_REG_HIPCIDR_BUSY) ||
+	    (done & CNL_DSP_REG_HIPCIDA_DONE))
```
Comment (keyonjie) [inline sound/soc/sof/intel/cnl.c]: (done & CNL_DSP_REG_HIPCIDA_DONE) means msg done, it is actually ready, you are using it opposite?

File: sound/soc/sof/intel/hda-ipc.c [hunk 2]
```diff
@@ -49,11 +49,13 @@ int hda_dsp_ipc_cmd_done(struct snd_sof_dev *sdev, int dir)
...
-	u64 val;
+	u64 busy, done;
 
 	/* is DSP ready for next IPC command */
-	val = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_HIPCI);
-	if (val & HDA_DSP_REG_HIPCI_BUSY)
+	busy = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_HIPCI);
+	done = snd_sof_dsp_read(sdev, HDA_DSP_BAR, HDA_DSP_REG_HIPCIE);
+	if ((busy & HDA_DSP_REG_HIPCI_BUSY) ||
+	    (done & HDA_DSP_REG_HIPCIE_DONE))
```
Comment (keyonjie) [inline sound/soc/sof/intel/hda-ipc.c]: ditto.

File: sound/soc/sof/intel/cnl.c (hunk 1 above)
Comment (RanderWang) [reply sound/soc/sof/intel/cnl.c]: no, it is right.  After DONE bit is set, the reture value of read would be zero.
if it is not set, the return value is 1. So it is different with we think.

**Comment by RanderWang** [comment]
After discussing with Keyon, we decide to unify all platforms in SOF to the same style. And this would kill some potential bugs.

Test on all the platform, it is easy to find the bug is fixed on APL CNL. And no regression issue on BYT, BDW. I dont have HSW, but it is almost the same as BDW.

File: sound/soc/sof/intel/hda-ipc.c (hunk 2 above)
Comment (RanderWang) [reply sound/soc/sof/intel/hda-ipc.c]: please refer to above reply

**Comment by lgirdwood** [comment]
@RanderWang I've fixed all sparse warnings meaning there are some conflicts with your PR, can you fix. Thanks

**Pull request closed without merging**